tests/0094-idempotence_msg_timeout.c

```diff
@@ -213,6 +215,114 @@ static void do_test_produce_timeout(const char *topic, const int msgrate) {
+        rd_kafka_t *rk;
+        rd_kafka_topic_t *rkt;
+        rd_kafka_conf_t *conf;
+        rd_kafka_pid_t pid, new_pid;
+        rd_kafka_resp_err_t err;
+        sockem_ctrl_t ctrl;
+        uint64_t testid;
+        char buf[1024];
+        char key[128];
+        int msgcnt          = 20;
+        const int partition = RD_KAFKA_PARTITION_UA;
+
+        SUB_TEST("Test idempotent producer epoch exhaustion");
+
+        testid = test_id_generate();
+
+        test_conf_init(&conf, NULL, 60);
+        sockem_ctrl_init(&ctrl);
+
+        test_conf_set(conf, "enable.idempotence", "true");
+        test_conf_set(conf, "batch.size", "1");
+        test_conf_set(conf, "linger.ms", "1");
+        test_conf_set(conf, "retries", "1");
+
+        rd_kafka_conf_set_dr_msg_cb(conf, my_dr_msg_cb);
+
+        test_socket_enable(conf);
+        test_curr->is_fatal_cb = is_fatal_cb;
+
+        rk  = test_create_handle(RD_KAFKA_PRODUCER, conf);
+        rkt = test_create_producer_topic(rk, topic, "message.timeout.ms", "100",
+                                         NULL);
+
+        /* Create the topic to make sure connections are up and ready. */
+        err = test_auto_create_topic_rkt(rk, rkt, tmout_multip(5000));
+        TEST_ASSERT(!err, "topic creation failed: %s", rd_kafka_err2str(err));
+
+        if (rd_kafka_test_idemp_get_pid(rk, &pid.id, &pid.epoch) != 0) {
+                TEST_FAIL(
+                    "rd_kafka_test_idemp_get_pid() failed for "
+                    "current PID");
+        }
+
+        /* Mock epoch near exhaustion point to make overflow faster */
+        if (rd_kafka_test_idemp_set_pid(rk, pid.id, INT16_MAX) != 0) {
+                TEST_FAIL("rd_kafka_test_idemp_set_pid() failed");
+        }
+
+        /* Add delay to socket to create message timeouts */
+        sockem_ctrl_set_delay(&ctrl, 500, 3 * 100);
+
+        for (int i = 0; i < msgcnt; i++) {
+                test_prepare_msg(testid, partition, i, buf, sizeof(buf), key,
+                                 sizeof(key));
+                err = rd_kafka_produce(rkt, partition, 0, buf, sizeof(buf), key,
+                                       sizeof(key), NULL);
+                if (err) {
+                        TEST_FAIL("produce() failed at msg #%d/%d: %s", i,
+                                  msgcnt, rd_kafka_err2str(err));
+                }
+
+                rd_usleep(100 * 1000, NULL);
+        }
+
+        /*  Remove delay. */
+        sockem_ctrl_set_delay(&ctrl, 0, 0);
+
+        test_flush(rk, 1000);
+
+        /* Check if we got some timeouted messages */
+        if (rd_kafka_test_idemp_get_pid(rk, &new_pid.id, &new_pid.epoch) != 0) {
+                TEST_FAIL(
+                    "rd_kafka_test_idemp_get_pid() failed for "
+                    "a new PID");
+        }
+        if (new_pid.epoch >= pid.epoch && new_pid.id == pid.id) {
+                TEST_FAIL("Expected Producer epoch exhaustion");
```

Comment:
nit: I find it easier to read TEST_ASSERT()s, e.g. what we are expecting:
`TEST_ASSERT(new_pid != pid, "Expected PID&Epoch bump");`